veps/sig-compute/hypervisor-abstraction.md

```diff
@@ -0,0 +1,588 @@
+
+- **Validation webhooks (`pkg/virt-api/webhooks/validating-webhook/admitters/hypervisor/`)** – We introduce a Validator interface that will define validation functions for core KubeVirt resources that have hypervisor-specific constraints, namely VM and VMI. Each hypervisor will provide its own concrete Validator to enforce rules and constraints relevant to its capabilities.
+
+  ```go
+  type Validator interface {
+      // Validate spec of VirtualMachine
+      ValidateVirtualMachineSpec(field *k8sfield.Path, spec *v1.VirtualMachineSpec, config *virtconfig.ClusterConfig) []metav1.StatusCause
+
+      // Validate spec of VirtualMachineInstance
+      ValidateVirtualMachineInstanceSpec(field *k8sfield.Path, spec *v1.VirtualMachineInstanceSpec, config *virtconfig.ClusterConfig) []metav1.StatusCause
+      
+      // Validate hot-plug updates to VMI. For example, this would encapsulate functionality in the ValidateHotplugDiskConfiguration function.
+      ValidateHotplug(oldVmi *v1.VirtualMachineInstance, newVmi *v1.VirtualMachineInstance, cc *virtconfig.ClusterConfig) []metav1.StatusCause
+  }
+  ```
+
+ - **Defaults provider registry (`pkg/defaults/providers/`)** – Introduces a single `DefaultsProvider` interface applied in layered order (Base → Hypervisor → Architecture → Hypervisor+Architecture → Finalization). Providers are registered under composite keys like `kvm/amd64` or `mshv/arm64`. Each provider implements:
+   ```go
+    type DefaultsProvider interface {
+        SetVirtualMachineDefaults(vm *v1.VirtualMachine, clusterConfig *virtconfig.ClusterConfig, virtClient kubecli.KubevirtClient)
+        SetDefaultVirtualMachineInstance(clusterConfig *virtconfig.ClusterConfig, vmi *v1.VirtualMachineInstance) error
+        SetDefaultVirtualMachineInstanceSpec(clusterConfig *virtconfig.ClusterConfig, spec *v1.VirtualMachineInstanceSpec) error
+    }
+   ```
+   Only zero-value fields are set at each layer; `FinalizeVMI` handles derived/status data (CPU topology snapshot, memory status, hotplug sizing, feature dependency resolution). Existing public functions delegate to the resolved provider for backwards compatibility.
+- **Runtime interfaces (`pkg/hypervisor/runtime/`)** – Provides interfaces for tuning a running VM or interacting with the virtualization stack outside of Libvirt. 
+  1.  `DomainTuner` interface for tuning a running virtual machine.
+      ```go
+      type DomainTuner interface {
+          // Adjust memlock limit for QEMU process
+          AdjustResources(vmi *v1.VirtualMachineInstance, additionalOverheadRatio *string) error
+          // Assign housekeeping CPU threads to housekeeping cgroup
+          HandleHousekeeping(vmi *v1.VirtualMachineInstance, domain *api.Domain) error
+      }
+      ```
+
+  2. `HypervisorCapabilityExtractor` interface for querying the hypervisor device for its capabilities.
+      ```go
+      type HypervisorCapabilityExtractor interface {
```

Comment:
Sorry for [circling back](https://github.com/kubevirt/enhancements/pull/98/files#r2541431389) to the naming topic, but I truly feel that we can save significant amounts of confusion in the future if we come up with concrete terms now for the different layers, especially since this will eventually become a public external API.

I think we need to come up with something like:
* KVM/MSHV layer - "Accelerator".
* QEMU/Cloud hypervisor - "Emulator".
* Libvirt - "Hypervisor".

We can discuss the different options, and we can do that in a follow-up PR to unblock the main design, but I think it would be much easier to settle this early in the process.